File: src/lib/api/justBoot.ts
import { invoke } from '@tauri-apps/api/core';

export interface JustBootEntry {
  id: string;
  ecid: string;
  productType: string;
  deviceName: string | null;
  buildId: string;
  iosVersion: string | null;
  bootArgs: string | null;
  repackedIbssPath: string | null;
  repackedIbecPath: string | null;
  sourceIpswPath: string | null;
  createdAt: string;
  lastBootedAt: string;
}

export interface JustBootEntryInput {
  ecid: string;
  productType: string;
  deviceName?: string | null;
  buildId: string;
  iosVersion?: string | null;
  bootArgs?: string | null;
  repackedIbssPath?: string | null;
  repackedIbecPath?: string | null;
  sourceIpswPath?: string | null;
}

export interface PrepareAndJustBootRequest {
  ecid: string;
  productType: string;
  deviceName?: string | null;
  buildId: string;
  iosVersion?: string | null;
  ipswPath: string;
  bootArgs?: string | null;
  processorGeneration?: string | null;
  includeIbec: boolean;
}

export function listJustBootHistory(): Promise<JustBootEntry[]> {
  return invoke<JustBootEntry[]>('list_just_boot_history');
}

export function recordJustBoot(entry: JustBootEntryInput): Promise<JustBootEntry> {
  return invoke<JustBootEntry>('record_just_boot', { entry });
}

export function forgetJustBoot(id: string): Promise<void> {
  return invoke<void>('forget_just_boot', { id });
}

export function prepareAndJustBoot(request: PrepareAndJustBootRequest): Promise<JustBootEntry> {
  return invoke<JustBootEntry>('prepare_and_just_boot', { request });
}

File: legacykit-ui/src-tauri/src/models/just_boot.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JustBootEntry {
    pub id: String,
    pub ecid: String,
    pub product_type: String,
    pub device_name: Option<String>,
    pub build_id: String,
    pub ios_version: Option<String>,
    pub boot_args: Option<String>,
    pub repacked_ibss_path: Option<String>,
    pub repacked_ibec_path: Option<String>,
    pub source_ipsw_path: Option<String>,
    pub created_at: String,
    pub last_booted_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JustBootEntryInput {
    pub ecid: String,
    pub product_type: String,
    pub device_name: Option<String>,
    pub build_id: String,
    pub ios_version: Option<String>,
    pub boot_args: Option<String>,
    pub repacked_ibss_path: Option<String>,
    pub repacked_ibec_path: Option<String>,
    pub source_ipsw_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareAndJustBootRequest {
    pub ecid: String,
    pub product_type: String,
    pub device_name: Option<String>,
    pub build_id: String,
    pub ios_version: Option<String>,
    pub ipsw_path: String,
    pub boot_args: Option<String>,
    pub processor_generation: Option<String>,
    pub include_ibec: bool,
}
